sharding/collator.py

```diff
@@ -0,0 +1,311 @@
+        main_chain_included=False):
+    """Verfiy the collation header only
+
+    Use for validate the ancestors or descendants collation header
+    """
+    try:
+        # The `coinbase` is a valid account
+        # TODO: check if the coinbase is in the same shard?
+        if not is_valid_account(collation_header.coinbase):
+            raise ValueError("coinbase is invalid")
+
+        # If `shardId >= 1 + SHARD_CHILD_COUNT + … + SHARD_CHILD_COUNT ** (MAX_SHARD_DEPTH-1)`, then the *children is an empty list*
+        # If `shardId < 1 + SHARD_CHILD_COUNT + … + SHARD_CHILD_COUNT ** (MAX_SHARD_DEPTH-1)`, then *each entry in the children list is either an empty string, or a hash whose preimage is available*, and whose shardId is this header’s shardId multiplied by SHARD_CHILD_COUNT plus (1 + the child’s index in the list)
+        verify_children_count(chain, children_header)
+        if len(children_header) > 0:
+            for index, child in enumerate(collation_header.children):
+                if child != '' and not is_valid_child(index, child, children_header):
+                    raise ValueError('child is wrong')
+
+        # `rng_source_block_number` is a block number equal to or greater than `parent_block_number`
+        if collation_header.rng_source_block_number < collation_header.parent_block_number:
+            raise ValueError('rng_source_block_number is lesser than parent_block_number')
+
+        # The `prev_state_root` is the current state root for the given shard
+        if collation_header.prev_state_root != prev_state.trie.root_hash:
+            raise ValueError("prev_state_root is wrong")
+
+        # Verify signatures
+        signature_result = verify_signature(chain, collation_header)
+        if signature_result is None and main_chain_included:
+            return False
+        elif signature_result is None and not main_chain_included:
+            return True
+        else:
+            return signature_result
+    except Exception as e:
+        print(str(e))
+        return False
+    return True
+
+
+def sign_collation_header(chain, collation_header, local_validation_code_addr):
+    try:
+        valid_signature_count = 0
+        for index, sig in enumerate(collation_header.signatures):
+            if len(sig) == 0:
+
+                validation_code_addr = contract_utils.call_sample_function(
+                    chain,
+                    block_number=collation_header.rng_source_block_number,
+                    shardId=chain.shardId,
+                    sigIndex=index)
+                if validation_code_addr == local_validation_code_addr:
+                    # the sig is empty
+                    # TODO: set collation_header.signatures[index] = local_sign
+                    collation_header.signatures[index] = b'\x11'
+                    valid_signature_count += 1
+                    break
+            else:
+                raise ValueError('signature is invalid')
```

Comment:
I pushed a new commit to simplify the `sign_collation_header and CollationHeader` function.
**TODO**: The optimization of times of calling `call_sample_function`